piet-scene/src/scene/builder.rs

```diff
@@ -51,44 +51,47 @@ impl<'a> SceneBuilder<'a> {
 
     /// Pushes a new layer bound by the specifed shape and composed with
     /// previous layers using the specified blend mode.
-    pub fn push_layer<'s, E>(&mut self, blend: BlendMode, transform: Affine, elements: E)
-    where
-        E: IntoIterator,
-        E::IntoIter: Clone,
-        E::Item: Borrow<PathElement>,
-    {
+    pub fn push_layer(
+        &mut self,
+        blend: impl Into<BlendMode>,
+        transform: Affine,
+        shape: &impl Shape,
+    ) {
+        let blend = blend.into();
         self.maybe_encode_transform(transform);
         self.linewidth(-1.0);
-        let elements = elements.into_iter();
-        self.encode_path(elements, true);
-        self.begin_clip(Some(blend));
-        self.layers.push(blend);
+        if self.encode_path(shape, true) {
+            self.begin_clip(blend);
+            self.layers.push((blend, true));
+        } else {
+            // When the layer has an empty path, record an entry to prevent
```

Comment:
I think I'm confused what the semantics should be for an empty path. Putting blends aside, clipping to an empty path should suppress rendering of the children altogether (not super useful, but it should do the right thing). Here it seems it renders the children without a clip.

For blends, the path given should enclose the content, so an empty path with nonempty content is a client bug. I'm less concerned about correct semantics in that case.